core/src/block_sync.rs

```diff
@@ -25,66 +32,103 @@ pub struct BlockSynchronizer {
     peer_id: PeerId,
     gossip_period: Duration,
     block_batch_size: u32,
-    broker: Broker,
-    actor_channel_capacity: u32,
-}
-
-#[async_trait::async_trait]
-impl Actor for BlockSynchronizer {
-    fn actor_channel_capacity(&self) -> u32 {
-        self.actor_channel_capacity
-    }
-
-    async fn on_start(&mut self, ctx: &mut Context<Self>) {
-        self.broker.subscribe::<message::Message, _>(ctx);
-        ctx.notify_every::<message::ReceiveUpdates>(self.gossip_period);
-    }
-}
-
-#[async_trait::async_trait]
-impl Handler<message::ReceiveUpdates> for BlockSynchronizer {
-    type Result = ();
-    async fn handle(&mut self, _: message::ReceiveUpdates) {
-        if let Some(random_peer) = self.sumeragi.get_random_peer_for_block_sync() {
-            self.request_latest_blocks_from_peer(random_peer.id.clone())
-                .await;
-        }
-    }
-}
-
-#[async_trait::async_trait]
-impl Handler<message::Message> for BlockSynchronizer {
-    type Result = ();
-    async fn handle(&mut self, message: message::Message) {
-        message.handle_message(self).await;
-    }
+    p2p: Arc<P2PSystem>,
 }
 
 impl BlockSynchronizer {
-    /// Sends request for latest blocks to a chosen peer
-    async fn request_latest_blocks_from_peer(&mut self, peer_id: PeerId) {
-        message::Message::GetBlocksAfter(message::GetBlocksAfter::new(
-            self.sumeragi.latest_block_hash(),
-            self.peer_id.clone(),
-        ))
-        .send_to(self.broker.clone(), peer_id)
-        .await;
-    }
-
     /// Create [`Self`] from [`Configuration`]
     pub fn from_configuration(
         config: &Configuration,
         sumeragi: Arc<Sumeragi>,
+        p2p: Arc<P2PSystem>,
         peer_id: PeerId,
-        broker: Broker,
-    ) -> Self {
-        Self {
+    ) -> Arc<Self> {
```

Comment:
Maybe it's a good idea to construct the `Arc` explicitly, and remove it from the signature.